### PR TITLE
Try fixing flutter hotrestart issues 

### PR DESCRIPTION
**Pull request by annagrin**
To perform the hot restart we need the app to be running, so we remove all the breakpoints, resume the app, then hot restart, and re-establish breakpoints after the main is reloaded. This seems to be too late for the breakpoints to hit.

Changes:

- Break hot restart into two steps - running the bootstrap and running main (flutter engine main in that case). 
- Run re-establishing the breakpoints can be done in between the two.
- Move hot restart and full reload logic from vm client to chrome proxy service to keep the proxy service state machine in one place.

Closes: https://github.com/dart-lang/webdev/issues/2018

### CODE REVIEW

File: dwds/lib/src/dwds_vm_client.dart [hunk 1]
```diff
@@ -147,14 +146,15 @@ class DwdsVmClient {
     return dwdsVmClient;
   }
 
-  Future<Map<String, dynamic>> hotRestart(
-      ChromeProxyService chromeProxyService, VmService client) {
-    return _hotRestartQueue.run(() => _hotRestart(chromeProxyService, client));
+/*
```
Comment (annagrin) [inline dwds/lib/src/dwds_vm_client.dart]: remove

File: dwds/lib/src/dwds_vm_client.dart [hunk 2]
```diff
@@ -193,17 +193,19 @@ void _recordDwdsStats(DwdsStats dwdsStats, String screen) {
   }
 }
 
+/*
```
Comment (annagrin) [inline dwds/lib/src/dwds_vm_client.dart]: remove

File: dwds/lib/src/services/chrome_proxy_service.dart [hunk 3]
```diff
@@ -520,8 +511,8 @@ ${globalLoadStrategy.loadModuleSnippet}("dart_sdk").developer.invokeExtension(
       final evaluator = _expressionEvaluator;
       if (evaluator != null) {
         await isCompilerInitialized;
-        _checkIsolate('evaluateInFrame', isolateId);
 
+        _checkIsolate('evaluateInFrame', isolateId);
```
Comment (annagrin) [inline dwds/lib/src/services/chrome_proxy_service.dart:515]: undo line changes

File: dwds/lib/src/services/chrome_proxy_service.dart [hunk 4]
```diff
@@ -1132,6 +1123,162 @@ ${globalLoadStrategy.loadModuleSnippet}("dart_sdk").developer.invokeExtension(
...
+  }
+
+  Future<void> _runMain() async {
+    // Wait for the debugger to get the first resume event and
+    // for the new inspector to be ready.
+    _logger.info('Waiting for the vm service to start.');
+    await isStarted;
+
+    _logger.info('Running dart main.');
+    await inspector.jsEvaluate('\$dartRunMain();');
```
Comment (annagrin) [inline dwds/lib/src/services/chrome_proxy_service.dart:1279]: awaitPromise: true ?

**Comment by mosuem** [comment]
Closing this, as it is stale and the files are out of date. Feel free to reopen in case you still want to land it!

**Pull request closed without merging**